### ISSUE TITLE
beforeHandler for NotFoundHandler

### ISSUE THREAD

**Issue by olnazx** [fastify/fastify#713]
[setNotFoundHandler](https://github.com/fastify/fastify/blob/master/docs/Server-Methods.md#setnotfoundhandler) description says
> The handler is treated like a regular route handler so requests will go through the full Fastify lifecycle.  

But actually `beforeHandler` is missing in the lifecycle, because there is no way to set it.

**Comment by delvedor** [comment]
The `beforeHandler` is an option that you can add to a specific route.
If you have two routes, A and B, and only B has the `beforeHandler`, A will not get into it.
The *before* handler technically is not part of the lifecycle since works on a single route level.

**Comment by nwoltman** [comment]
@olnazx Is there any reason why you can't put your `beforeHandler` logic inside your not-found handler? While the not-found handler is similar to a route, it's a special case and currently does not allow setting the options that a normal route does.

**Comment by olnazx** [comment]
@nwoltman the main reason is that it leads to code duplication. I have several not-found handlers and I have to put the same `beforeHandler` into each of them. 

As I see `setNotFoundHandler` accepts options object as a first argument (not mentioned in the docs though). I'm not really sure why it's not allowed to set `beforeHandler` option (maybe it will break something), but implementation of `beforeHandler` for not-found handler as in a regular route seems to work for me. I mean [L659](https://github.com/fastify/fastify/blob/master/fastify.js#L659)
```js
const preHandler = this._hooks.preHandler.concat(opts.beforeHandler || [])
```

**Comment by mcollina** [comment]
I see no harm in adding this. Would you like to send a PR @olnazx?

**Comment by israeleriston** [comment]
@mcollina will addin into documentation to set NotFoundHandler or will implemented suggestion the @olnazx ?

**Comment by mcollina** [comment]
It would be awesome to get a PR for this!

**Comment by israeleriston** [comment]
I understood @mcollina , I just wanted to understand, what needs to be done.

**Comment by mcollina** [comment]
Add a beforeHandler option to the setNotFoundHandler. Or I have not understood the question.

**Comment by israeleriston** [comment]
I'll get a PR for this. I just need a confirmation the @olnazx , if that's what it's supposed to do.

**Comment by nwoltman** [comment]
This issue should probably stay open. It is a very reasonable feature request. It just needs someone to make a PR.

**Comment by lependu** [comment]
If nobody works on it, I am happy to create a PR.

**Comment by mcollina** [comment]
Go for it!
Il giorno ven 29 giu 2018 alle 13:29 Akos Kovacs <notifications@github.com>
ha scritto:

**Comment by delvedor** [comment]
Closed in #1005
